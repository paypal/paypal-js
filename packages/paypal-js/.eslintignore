dist/*
scripts/*
v6-docs/*
