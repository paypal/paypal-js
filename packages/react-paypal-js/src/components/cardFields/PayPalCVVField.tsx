import React from "react";

import { PayPalCardFieldsIndividualFieldOptions } from "../../types";
import { PayPalCardField } from "./PayPalCardField";

export const PayPalCVVField: React.FC<
    PayPalCardFieldsIndividualFieldOptions
> = (options) => {
    return <PayPalCardField fieldName="CVVField" {...options} />;
};
