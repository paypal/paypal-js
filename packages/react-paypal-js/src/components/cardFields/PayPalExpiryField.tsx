import React from "react";

import { PayPalCardFieldsIndividualFieldOptions } from "../../types";
import { PayPalCardField } from "./PayPalCardField";

export const PayPalExpiryField: React.FC<
    PayPalCardFieldsIndividualFieldOptions
> = (options) => {
    return <PayPalCardField fieldName="ExpiryField" {...options} />;
};
