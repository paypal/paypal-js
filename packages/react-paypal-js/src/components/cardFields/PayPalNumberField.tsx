import React from "react";

import { PayPalCardFieldsIndividualFieldOptions } from "../../types";
import { PayPalCardField } from "./PayPalCardField";

export const PayPalNumberField: React.FC<
    PayPalCardFieldsIndividualFieldOptions
> = (options) => {
    return <PayPalCardField fieldName="NumberField" {...options} />;
};
