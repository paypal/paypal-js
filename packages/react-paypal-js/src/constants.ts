/*********************************************
 * Common reference to the script identifier *
 *********************************************/

// keep this script id value in kebab-case format
export const SCRIPT_ID = "data-react-paypal-script-id";

export const SDK_SETTINGS = {
    DATA_CLIENT_TOKEN: "dataClientToken",
    DATA_JS_SDK_LIBRARY: "dataJsSdkLibrary",
    DATA_LIBRARY_VALUE: "react-paypal-js",
    DATA_NAMESPACE: "dataNamespace",
    DATA_SDK_INTEGRATION_SOURCE: "dataSdkIntegrationSource",
    DATA_USER_ID_TOKEN: "dataUserIdToken",
};
export const LOAD_SCRIPT_ERROR = "Failed to load the PayPal JS SDK script.";

/****************************
 * Braintree error messages *
 ****************************/
export const EMPTY_BRAINTREE_AUTHORIZATION_ERROR_MESSAGE =
    "Invalid authorization data. Use dataClientToken or dataUserIdToken to authorize.";

const braintreeVersion = "3.84.0";
export const BRAINTREE_SOURCE = `https://js.braintreegateway.com/web/${braintreeVersion}/js/client.min.js`;
export const BRAINTREE_PAYPAL_CHECKOUT_SOURCE = `https://js.braintreegateway.com/web/${braintreeVersion}/js/paypal-checkout.min.js`;

/*********************
 * PayPal namespaces *
 *********************/
export const DEFAULT_PAYPAL_NAMESPACE = "paypal";
export const DEFAULT_BRAINTREE_NAMESPACE = "braintree";

/*****************
 * Hosted Fields *
 *****************/
export const HOSTED_FIELDS_CHILDREN_ERROR =
    "To use HostedFields you must use it with at least 3 children with types: [number, cvv, expirationDate] includes";
export const HOSTED_FIELDS_DUPLICATE_CHILDREN_ERROR =
    "Cannot use duplicate HostedFields as children";

/*******************
 * Script Provider *
 *******************/
export const SCRIPT_PROVIDER_REDUCER_ERROR =
    "usePayPalScriptReducer must be used within a PayPalScriptProvider";

/*******************
 * Card Fields *
 *******************/
export const CARD_ERRORS = {
    INELIGIBLE_CARD_VENDOR: "INELIGIBLE_CARD_VENDOR",
    INVALID_NUMBER: "INVALID_NUMBER",
    INVALID_EXPIRY: "INVALID_EXPIRY",
    INVALID_CVV: "INVALID_CVV",
    INVALID_NAME: "INVALID_NAME",
} as const;
export const CARD_FIELDS_CHILDREN_ERROR =
    "To use CardFields you must use it with at least the 3 required fields (PayPalNumberField, PayPalCVVField, PayPalExpiryField)";
export const CARD_FIELDS_DUPLICATE_CHILDREN_ERROR =
    "Cannot use duplicate CardFields as children";
export const CARD_FIELDS_CONTEXT_ERROR =
    "Individual CardFields must be rendered inside the PayPalCardFieldsProvider";
