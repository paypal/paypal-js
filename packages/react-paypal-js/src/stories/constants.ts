import { FUNDING, CURRENCY } from "@paypal/sdk-constants/dist/module";

export const COMPONENT_PROPS_CATEGORY = { table: { category: "Props" } };
export const COMPONENT_EVENTS = "Events";
export const COMPONENT_TYPES = "Types";
export const ORDER_ID = "orderID";
export const ERROR = "Error";
export const SDK = "SDK";
export const APPROVE = "approve";
export const SUBSCRIPTION = "subscription";
export const ORDER_INSTANCE_ERROR = "No order instance was provided";

export const CONTAINER_SIZE = {
    name: "container width",
    description:
        "This is not a property from PayPalButtons. It is custom control to change the size of the PayPal buttons container in pixels.",
    control: { type: "range", min: 200, max: 750, step: 5 },
    table: {
        defaultValue: {
            summary: "750px",
        },
        category: "Custom",
        type: { summary: "number" },
    },
};

export const ARG_TYPE_AMOUNT = {
    description:
        "This is not a property from PayPalButtons. It is custom control for testing the amount sent in the createOrder process",
    options: ["2", "30", "100"],
    control: {
        type: "select",
    },
    table: {
        defaultValue: {
            summary: "2.00",
        },
        category: "Custom",
        type: { summary: "number|string" },
    },
};

export const ARG_TYPE_CURRENCY = {
    options: Object.values(CURRENCY),
    description:
        "This is not a property from PayPalButtons. It is custom control to change the currency create create a PayPal order.",
    control: {
        type: "select",
    },
    table: {
        category: "Custom",
        type: { summary: "string" },
        defaultValue: { summary: "USD" },
    },
};

export const FUNDING_SOURCE_ARG = {
    options: [...Object.values(FUNDING), undefined],
    control: {
        type: "select",
        labels: {
            undefined: "all",
        },
    },
    ...COMPONENT_PROPS_CATEGORY,
};

export const SANDPACK_STYLES = `body {
    font-family: sans-serif;
    -webkit-font-smoothing: auto;
    -moz-font-smoothing: auto;
    -moz-osx-font-smoothing: grayscale;
    font-smoothing: auto;
    text-rendering: optimizeLegibility;
    font-smooth: always;
    -webkit-tap-highlight-color: transparent;
    -webkit-touch-callout: none;
}

.card-field {
    width: 100%;
    padding: 12px;
    border: 1px solid #ccc;
    border-radius: 4px;
    box-sizing: border-box;
    margin-top: 6px;
    margin-bottom: 16px;
    resize: vertical;
    height: 40px;
    background: white;
    font-size: 17px;
    color: #3a3a3a;
    font-family: helvetica, tahoma, calibri, sans-serif;
}

.btn {
    display: inline-block;
    font-weight: 400;
    text-align: center;
    white-space: nowrap;
    vertical-align: middle;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    border: 1px solid transparent;
    padding: 0.375rem 0.75rem;
    font-size: 1rem;
    line-height: 1.5;
    border-radius: 0.25rem;
    transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
        border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
    cursor: pointer;
}

.btn-primary {
    color: #fff;
    background-color: #007bff;
    border-color: #007bff;
}

.mark {
    display: flex;
    align-items: center;
}
.spinner {
    display: inline-block;
    height: 40px;
    width: 40px;
    box-sizing: border-box;
    border: 3px solid rgba(0, 0, 0, 0.2);
    border-top-color: rgba(33, 128, 192, 0.8);
    border-radius: 100%;
    animation: rotation 0.7s infinite linear;
}

.spinner.tiny {
    height: 20px;
    width: 20px;
    border-top-color: #007bff;
    position: relative;
    top: 3px;
}

@keyframes rotation {
    0% {
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(359deg);
    }
}`;
