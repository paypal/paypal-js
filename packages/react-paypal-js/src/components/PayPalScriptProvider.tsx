import React, { useEffect, useReducer } from "react";
import { loadScript } from "@paypal/paypal-js";

import {
    getScriptID,
    ScriptContext,
    scriptReducer,
} from "../context/scriptProviderContext";
import { SCRIPT_ID, SDK_SETTINGS, LOAD_SCRIPT_ERROR } from "../constants";
import { SCRIPT_LOADING_STATE, DISPATCH_ACTION } from "../types";

import type { FC } from "react";
import type { ScriptProviderProps } from "../types";

/**
This `<PayPalScriptProvider />` component takes care of loading the JS SDK `<script>`.
It manages state for script loading so children components like `<PayPalButtons />` know when it's safe to use the `window.paypal` global namespace.

Note: You always should use this component as a wrapper for  `PayPalButtons`, `PayPalMarks`, `PayPalMessages` and `BraintreePayPalButtons` components.
 */
export const PayPalScriptProvider: FC<ScriptProviderProps> = ({
    options = { clientId: "test" },
    children,
    deferLoading = false,
}: ScriptProviderProps) => {
    const [state, dispatch] = useReducer(scriptReducer, {
        options: {
            ...options,
            [SDK_SETTINGS.DATA_JS_SDK_LIBRARY]: SDK_SETTINGS.DATA_LIBRARY_VALUE,
            [SDK_SETTINGS.DATA_SDK_INTEGRATION_SOURCE]:
                SDK_SETTINGS.DATA_LIBRARY_VALUE,
            [SCRIPT_ID]: `${getScriptID(options)}`,
        },
        loadingStatus: deferLoading
            ? SCRIPT_LOADING_STATE.INITIAL
            : SCRIPT_LOADING_STATE.PENDING,
    });

    useEffect(() => {
        if (
            deferLoading === false &&
            state.loadingStatus === SCRIPT_LOADING_STATE.INITIAL
        ) {
            return dispatch({
                type: DISPATCH_ACTION.LOADING_STATUS,
                value: SCRIPT_LOADING_STATE.PENDING,
            });
        }

        if (state.loadingStatus !== SCRIPT_LOADING_STATE.PENDING) {
            return;
        }

        let isSubscribed = true;

        loadScript(state.options)
            .then(() => {
                if (isSubscribed) {
                    dispatch({
                        type: DISPATCH_ACTION.LOADING_STATUS,
                        value: SCRIPT_LOADING_STATE.RESOLVED,
                    });
                }
            })
            .catch((err) => {
                console.error(`${LOAD_SCRIPT_ERROR} ${err}`);
                if (isSubscribed) {
                    dispatch({
                        type: DISPATCH_ACTION.LOADING_STATUS,
                        value: {
                            state: SCRIPT_LOADING_STATE.REJECTED,
                            message: String(err),
                        },
                    });
                }
            });
        return () => {
            isSubscribed = false;
        };
    }, [state.options, deferLoading, state.loadingStatus]);

    return (
        <ScriptContext.Provider value={{ ...state, dispatch }}>
            {children}
        </ScriptContext.Provider>
    );
};
