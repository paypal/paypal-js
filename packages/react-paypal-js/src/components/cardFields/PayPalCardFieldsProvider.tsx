import React, { ReactNode, useEffect, useRef, useState } from "react";

import { usePayPalScriptReducer } from "../../hooks/scriptProviderHooks";
import { getPayPalWindowNamespace } from "../../utils";
import { SDK_SETTINGS } from "../../constants";
import { generateMissingCardFieldsError } from "./utils";
import { PayPalCardFieldsContext } from "./context";
import { usePayPalCardFieldsRegistry } from "./hooks";
import { FullWidthContainer } from "../ui/FullWidthContainer";

import type {
    PayPalCardFieldsComponentOptions,
    PayPalCardFieldsComponent,
} from "@paypal/paypal-js/types/components/card-fields";

type CardFieldsProviderProps = PayPalCardFieldsComponentOptions & {
    children: ReactNode;
};

/**
The `<PayPalCardFieldsProvider />` is a context provider that is designed to support the rendering and state management of PayPal CardFields in your application. 

The context provider will initialize the `CardFields` instance from the JS SDK and determine eligibility to render the CardField components. Once the `CardFields` are initialized, the context provider will manage the state of the `CardFields` instance as well as the reference to each individual card field.

Passing the `inputEvents` and `style` props to the context provider will apply them to each of the individual field components.

The state managed by the provider is accessible through our custom hook `usePayPalCardFields`.

*/

export const PayPalCardFieldsProvider = ({
    children,
    ...props
}: CardFieldsProviderProps): JSX.Element => {
    const [{ isResolved, options }] = usePayPalScriptReducer();
    const { fields, registerField, unregisterField } =
        usePayPalCardFieldsRegistry();

    const [cardFieldsForm, setCardFieldsForm] =
        useState<PayPalCardFieldsComponent | null>(null);
    const cardFieldsInstance = useRef<PayPalCardFieldsComponent | null>(null);

    const [isEligible, setIsEligible] = useState(false);
    // We set the error inside state so that it can be caught by React's error boundary
    const [, setError] = useState(null);

    useEffect(() => {
        if (!isResolved) {
            return;
        }

        try {
            cardFieldsInstance.current =
                getPayPalWindowNamespace(
                    options[SDK_SETTINGS.DATA_NAMESPACE],
                ).CardFields?.({
                    ...props,
                }) ?? null;
        } catch (error) {
            setError(() => {
                throw new Error(
                    `Failed to render <PayPalCardFieldsProvider /> component. Failed to initialize:  ${error}`,
                );
            });
            return;
        }

        if (!cardFieldsInstance.current) {
            setError(() => {
                throw new Error(
                    generateMissingCardFieldsError({
                        components: options.components,
                        [SDK_SETTINGS.DATA_NAMESPACE]:
                            options[SDK_SETTINGS.DATA_NAMESPACE],
                    }),
                );
            });
            return;
        }

        setIsEligible(cardFieldsInstance.current.isEligible());
        setCardFieldsForm(cardFieldsInstance.current);

        return () => {
            setCardFieldsForm(null);
            cardFieldsInstance.current = null;
        };
    }, [isResolved]); // eslint-disable-line react-hooks/exhaustive-deps

    if (!isEligible) {
        // TODO: What should be returned here?
        return <div />;
    }

    return (
        <FullWidthContainer>
            <PayPalCardFieldsContext.Provider
                value={{
                    cardFieldsForm,
                    fields,
                    registerField,
                    unregisterField,
                }}
            >
                {children}
            </PayPalCardFieldsContext.Provider>
        </FullWidthContainer>
    );
};
