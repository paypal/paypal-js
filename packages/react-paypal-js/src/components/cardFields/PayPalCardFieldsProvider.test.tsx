import React from "react";
import { render, waitFor } from "@testing-library/react";
import { ErrorBoundary } from "react-error-boundary";
import { PayPalNamespace, loadScript } from "@paypal/paypal-js";
import { mock } from "jest-mock-extended";

import { PayPalScriptProvider } from "../PayPalScriptProvider";
import { PayPalCardFieldsProvider } from "./PayPalCardFieldsProvider";
import { CARD_FIELDS_DUPLICATE_CHILDREN_ERROR } from "../../constants";
import { PayPalNumberField } from "./PayPalNumberField";
import { PayPalCVVField } from "./PayPalCVVField";
import { PayPalExpiryField } from "./PayPalExpiryField";

import type { PayPalCardFieldsComponent } from "@paypal/paypal-js/types/components/card-fields";
import type { ReactNode } from "react";

const MOCK_ELEMENT_ID = "mock-element";

jest.mock("@paypal/paypal-js", () => ({
    loadScript: jest.fn(),
}));

function getMockElementsRendered() {
    return document.querySelectorAll(`div[id^=${MOCK_ELEMENT_ID}]`);
}

const mockCreateOrder = mock<() => Promise<string>>();
const mockOnApprove = mock<() => Promise<string>>();
const mockOnError = mock<() => void>();
const onError = jest.fn();

const wrapper = ({ children }: { children: ReactNode }) => (
    <ErrorBoundary fallback={<div>Error</div>} onError={onError}>
        {children}
    </ErrorBoundary>
);

const isEligible = jest.fn();

const happyMethods = {
    render: jest.fn((element: string | HTMLElement) => {
        // Simulate adding element
        if (typeof element !== "string") {
            const child = document.createElement("div");
            child.setAttribute("id", MOCK_ELEMENT_ID);
            element.append(child);
        }
        return Promise.resolve();
    }),
    close: jest.fn(() => Promise.resolve()),
};
const unHappyMethods = {
    render: jest.fn((element: string | HTMLElement) => {
        // Simulate adding element
        if (typeof element !== "string") {
            element.append(document.createElement("div"));
        }
        return Promise.reject("Unknown error");
    }),
    close: jest.fn(() => Promise.reject("Unknown error")),
};

describe("PayPalCardFieldsProvider", () => {
    beforeEach(() => {
        document.body.innerHTML = "";

        window.paypal = {
            CardFields: jest.fn(
                () =>
                    ({
                        isEligible: isEligible.mockReturnValue(true),
                        NumberField: jest.fn().mockReturnValue(happyMethods),
                        NameField: jest.fn().mockReturnValue(happyMethods),
                        CVVField: jest.fn().mockReturnValue(happyMethods),
                        ExpiryField: jest.fn().mockReturnValue(happyMethods),
                    } as unknown as PayPalCardFieldsComponent)
            ),
            version: "",
        };

        (loadScript as jest.Mock).mockResolvedValue(window.paypal);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test("should throw an Error using the component without the PayPalScriptProvider", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        render(
            <PayPalCardFieldsProvider
                onApprove={mockOnApprove}
                createOrder={mockCreateOrder}
                onError={mockOnError}
            >
                <></>
            </PayPalCardFieldsProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toEqual(
            "usePayPalScriptReducer must be used within a PayPalScriptProvider"
        );
        spyConsoleError.mockRestore();
    });

    test("should throw an Error using the component with duplicate fields", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test-client",
                    currency: "USD",
                    intent: "authorize",
                    components: "card-fields",
                    dataClientToken: "test-data-client-token",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                    <PayPalNumberField />
                    <PayPalCVVField />
                    <PayPalExpiryField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toEqual(
            CARD_FIELDS_DUPLICATE_CHILDREN_ERROR
        );

        spyConsoleError.mockRestore();
    });

    test("should return immediately when script provider is rejected", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        (loadScript as jest.Mock).mockRejectedValue(new Error("Unknown error"));

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test-client",
                    currency: "USD",
                    intent: "authorize",
                    components: "card-fields",
                    dataClientToken: "test-data-client-token",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                    <PayPalCVVField />
                    <PayPalExpiryField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>
        );
        await waitFor(() => {
            expect(loadScript).toBeCalled();
        });
        expect(getMockElementsRendered().length).toEqual(0);
        expect(onError).toBeCalledTimes(0);

        spyConsoleError.mockRestore();
    });

    test("should not render CardFields components when unilegible", async () => {
        window.paypal = {
            CardFields: jest.fn(
                () =>
                    ({
                        isEligible: jest.fn().mockReturnValue(false),
                    } as unknown as PayPalCardFieldsComponent)
            ),
            version: "",
        };

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test-client",
                    currency: "USD",
                    intent: "authorize",
                    components: "card-fields",
                    dataClientToken: "test-data-client-token",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                    <PayPalCVVField />
                    <PayPalExpiryField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>
        );
        await waitFor(() => {
            expect(getMockElementsRendered().length).toEqual(0);
        });
    });

    test("should catch and throw unexpected zoid render errors", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        window.paypal = {
            CardFields: jest.fn(
                () =>
                    ({
                        isEligible: jest.fn().mockReturnValue(true),
                        NumberField: jest.fn(() => unHappyMethods),
                    } as unknown as PayPalCardFieldsComponent)
            ),
            version: "",
        };

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test-client",
                    currency: "USD",
                    intent: "authorize",
                    dataClientToken: "test-data-client-token",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toEqual(
            "Failed to render <PayPalNumberField /> component. Unknown error"
        );
        spyConsoleError.mockRestore();
    });

    test("should throw an error when the 'card-fields' component is missing from the components list passed to the PayPalScriptProvider", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        window.paypal = { CardFields: undefined } as PayPalNamespace;

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test",
                    components: "buttons,messages",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                    <PayPalCVVField />
                    <PayPalExpiryField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toMatchSnapshot();
        spyConsoleError.mockRestore();
    });

    test("should render card fields", async () => {
        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test-client",
                    currency: "USD",
                    intent: "authorize",
                    dataClientToken: "test-data-client-token",
                    components: "card-fields",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                    <PayPalCVVField />
                    <PayPalExpiryField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>
        );

        await waitFor(() => {
            expect(getMockElementsRendered().length).toEqual(3);
        });
    });

    test("should safely ignore error on render process when paypal buttons container is no longer in the DOM ", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        const mockRender = jest
            .fn()
            .mockRejectedValue(new Error("Unknown error"));
        window.paypal = {
            CardFields: jest.fn(
                () =>
                    ({
                        isEligible: jest.fn().mockReturnValue(true),
                        NumberField: jest.fn(() => ({
                            render: mockRender,
                            close: happyMethods.close,
                        })),
                        CVVField: jest.fn(() => ({
                            render: mockRender,
                            close: happyMethods.close,
                        })),
                        ExpiryField: jest.fn(() => ({
                            render: mockRender,
                            close: happyMethods.close,
                        })),
                    } as unknown as PayPalCardFieldsComponent)
            ),
            version: "",
        };

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test-client",
                    currency: "USD",
                    intent: "authorize",
                    dataClientToken: "test-data-client-token",
                    components: "card-fields",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                    <PayPalCVVField />
                    <PayPalExpiryField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>
        );

        await waitFor(() => expect(mockRender).toHaveBeenCalledTimes(3));
        expect(onError).toBeCalledTimes(0);
        expect(getMockElementsRendered().length).toEqual(0);

        spyConsoleError.mockRestore();
    });
});
