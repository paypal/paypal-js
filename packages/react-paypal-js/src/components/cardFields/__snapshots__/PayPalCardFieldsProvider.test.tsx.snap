// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayPalCardFieldsProvider should throw an error when the 'card-fields' component is missing from the components list passed to the PayPalScriptProvider 1`] = `
"Unable to render <PayPalCardFieldsProvider /> because window.paypal.CardFields is undefined.
To fix the issue, add 'card-fields' to the list of components passed to the parent PayPalScriptProvider: <PayPalScriptProvider options={{ components: 'buttons,messages,card-fields'}}>"
`;
